File: pilots/schur-six/verify.py
"""Exact six-color Schur witness checker. Standard library only; no execution of artifacts."""
import argparse
import hashlib
import json
import math
from pathlib import Path

MAX_BYTES = 1024 * 1024
MAX_N = 10000


def _integer(value):
    return type(value) is int or (type(value) is float and math.isfinite(value) and value.is_integer())


def _object(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ValueError('Duplicate JSON key: ' + key)
        result[key] = value
    return result


def _constant(value):
    raise ValueError('Non-JSON numeric constant: ' + value)


def parse(data):
    if len(data) > MAX_BYTES:
        raise ValueError('Artifact exceeds 1 MiB')
    return json.loads(data.decode('utf-8'), object_pairs_hook=_object, parse_constant=_constant)


def verify(artifact):
    if type(artifact) is not dict or set(artifact) != {'format_version', 'n', 'colors'}:
        return {'valid': False, 'error': 'Expected exactly format_version, n, colors'}
    if not _integer(artifact['format_version']) or artifact['format_version'] != 1:
        return {'valid': False, 'error': 'Unsupported format_version'}
    n = artifact['n']
    if not _integer(n) or not 1 <= n <= MAX_N:
        return {'valid': False, 'error': 'n must be an integer from 1 to 10000'}
    n = int(n)
    colors = artifact['colors']
    if type(colors) is not list or len(colors) != n:
        return {'valid': False, 'error': 'colors must have exactly n entries'}
    if any(not _integer(c) or not 0 <= c <= 5 for c in colors):
        return {'valid': False, 'error': 'Each color must be an integer from 0 to 5; booleans are invalid'}
    colors = [int(c) for c in colors]
    checked = 0
    for x in range(1, n // 2 + 1):
        for y in range(x, n - x + 1):
            checked += 1
            z = x + y
            if colors[x - 1] == colors[y - 1] == colors[z - 1]:
                return {'valid': False, 'error': 'Monochromatic sum', 'counterexample': [x, y, z], 'color': colors[z - 1], 'pairs_checked': checked}
    return {'valid': True, 'n': n, 'colors_allowed': 6, 'class_sizes': [colors.count(c) for c in range(6)], 'pairs_checked': checked, 'claim': f'S(6) >= {n}'}


def verify_bytes(data):
    if len(data) > MAX_BYTES:
        return {'valid': False, 'error': 'Artifact exceeds 1 MiB'}
    try:
        result = verify(parse(data))
    except (ValueError, UnicodeError, RecursionError) as exc:
        result = {'valid': False, 'error': str(exc)}
    return dict(result, artifact_sha256=hashlib.sha256(data).hexdigest())


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('artifact', type=Path)
    args = parser.parse_args()
    try:
        with args.artifact.open('rb') as source:
            data = source.read(MAX_BYTES + 1)
    except OSError as exc:
        print(json.dumps({'valid': False, 'error': str(exc)}))
        return 2
    result = verify_bytes(data)
    print(json.dumps(result, sort_keys=True))
    return 0 if result['valid'] else 1


if __name__ == '__main__':
    raise SystemExit(main())

File: pilots/schur-six/verify.mjs
// Exact second implementation: bitset intersections, not the Python pair loop.
// Node.js standard library only. Does not execute artifact content.
import fs from 'node:fs';
import crypto from 'node:crypto';
import { pathToFileURL } from 'node:url';

export const MAX_BYTES = 1024 * 1024;
export const MAX_N = 10000;

export function parseArtifact(bytes) {
  if (bytes.length > MAX_BYTES) throw new Error('Artifact exceeds 1 MiB');
  const text = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }).decode(bytes);
  const value = JSON.parse(text);
  // JSON.parse otherwise keeps the last duplicate. Scan complete string tokens
  // at root depth so equivalent escaped keys are rejected too.
  let depth = 0;
  const keys = new Set();
  for (let i = 0; i < text.length;) {
    const ch = text[i];
    if (ch === '"') {
      const start = i++;
      while (i < text.length) {
        if (text[i] === '\\') { i += 2; continue; }
        if (text[i++] === '"') break;
      }
      if (depth === 1 && /^\s*:/.test(text.slice(i))) {
        const key = JSON.parse(text.slice(start, i));
        if (keys.has(key)) throw new Error('Duplicate JSON key: ' + key);
        keys.add(key);
      }
      continue;
    }
    if (ch === '{' || ch === '[') depth++;
    if (ch === '}' || ch === ']') depth--;
    i++;
  }
  return value;
}

export function verify(artifact) {
  if (artifact === null || Array.isArray(artifact) || typeof artifact !== 'object' || Object.keys(artifact).sort().join(',') !== 'colors,format_version,n')
    return { valid: false, error: 'Expected exactly format_version, n, colors' };
  if (!Number.isInteger(artifact.format_version) || artifact.format_version !== 1)
    return { valid: false, error: 'Unsupported format_version' };
  const { n, colors } = artifact;
  if (!Number.isInteger(n) || n < 1 || n > MAX_N)
    return { valid: false, error: 'n must be an integer from 1 to 10000' };
  if (!Array.isArray(colors) || colors.length !== n)
    return { valid: false, error: 'colors must have exactly n entries' };
  if (!colors.every(c => Number.isInteger(c) && c >= 0 && c <= 5))
    return { valid: false, error: 'Each color must be an integer from 0 to 5; booleans are invalid' };
  const masks = Array(6).fill(0n);
  const sizes = Array(6).fill(0);
  for (let i = 0; i < n; i++) {
    masks[colors[i]] |= 1n << BigInt(i + 1);
    sizes[colors[i]]++;
  }
  let shifts = 0;
  for (let x = 1; x <= n; x++) {
    const color = colors[x - 1];
    let overlap = (masks[color] << BigInt(x)) & masks[color];
    shifts++;
    if (overlap !== 0n) {
      let z = 0;
      while ((overlap & 1n) === 0n) { overlap >>= 1n; z++; }
      return { valid: false, error: 'Monochromatic sum', counterexample: [x, z - x, z], color, shifts_checked: shifts };
    }
  }
  return { valid: true, n, colors_allowed: 6, class_sizes: sizes, shifts_checked: shifts, claim: `S(6) >= ${n}` };
}

export function verifyBytes(bytes) {
  if (bytes.length > MAX_BYTES) return { valid: false, error: 'Artifact exceeds 1 MiB' };
  let result;
  try { result = verify(parseArtifact(bytes)); }
  catch (error) { result = { valid: false, error: String(error.message ?? error) }; }
  return { ...result, artifact_sha256: crypto.createHash('sha256').update(bytes).digest('hex') };
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  if (process.argv.length !== 3) { console.error('Usage: node verify.mjs artifact.json'); process.exitCode = 2; }
  else {
    let fd;
    try {
      fd = fs.openSync(process.argv[2], 'r');
      const bytes = Buffer.alloc(MAX_BYTES + 1);
      let used = 0;
      for (;;) {
        const got = fs.readSync(fd, bytes, used, bytes.length - used, null);
        used += got;
        if (!got || used === bytes.length) break;
      }
      const result = verifyBytes(bytes.subarray(0, used));
      console.log(JSON.stringify(result));
      process.exitCode = result.valid ? 0 : 1;
    } catch (error) { console.log(JSON.stringify({ valid: false, error: error.message })); process.exitCode = 2; }
    finally { if (fd !== undefined) fs.closeSync(fd); }
  }
}
